src/dmd/errors.d

```diff
@@ -847,3 +904,215 @@ private void writeHighlights(Console con, ref const OutBuffer buf)
             fputc(c, con.fp);
     }
 }
+
+/**
+ * Similar to `verrorPrint` except this function uses an extended format string syntax, and
+ * outsources the actual printing to external functions.
+ *
+ * Notes:
+ *  When this comment mentions about inserting a new line, what really happens it that is makes
+ *  a call to `printSupplemental` instead of manually messing around with new lines.
+ *
+ * Format:
+ *  The given `format` string can be used as a normal printf-style string, but it has additional syntax
+ *  to perform additional formatting options, such as indenting text depending on the error format level set by the compiler caller.
+ *
+ *  To specify a 'format fragment' you must use the syntax `$[formatter_name:formatter_params]{text to format}`.
+ *
+ *  formatter_params (and the ':' before it) are optional, everything else is mandatory.
+ *
+ *  e.g. `$[indent:1]{This text is put onto a new line and indented by 1 'tab'}`
+ *
+ *  e.g. `There might be a new line $[indent:0]{} between these two pieces of text!`
+ *
+ * Formatters:
+ *  indent:indent_level = If the error formatting level is 1 or higher, then a new line is made and a tab (4 spaces) is inserted `indent_level`
+ *                        times, which is then followed up by the text to format. Technically there's no need to provide specific text to this formatter,
+ *                        but it makes it more clear on what you're attempting to do.
+ *
+ * Params:
+ *  loc                 = The location of the error.
+ *  format              = The extended format string to use.
+ *  ap                  = The argument list to use with the `format` string.
+ *  print               = The print function to use for the first line printed.
+ *  printSupplemental   = The print function to use for every line printed after the first line.
+ */
+private void verrorFormatPrint(const ref Loc loc, const(char)* format, va_list ap, MessageFunc print, MessageFunc printSupplemental)
```

Comment:
... this could be templated to hide the `MessageFunc` type.
```suggestion
private void verrorFormatPrint(Fun)(const ref Loc loc, const(char)* format, va_list ap, Fun print, Fun printSupplemental)
```